pvlib/tracking.py

```diff
@@ -569,3 +582,163 @@ def singleaxis(apparent_zenith, apparent_azimuth,
+    rot = np.roll(rot, (axis, axis), (1, 0))
+    return rot
+
+
+def calc_tracker_axis_tilt(system_azimuth, system_zenith, tracker_azimuth):
+    """
+    Calculate tracker axis tilt in the global reference frame when on a sloped
+    plane.
+
+    Parameters
+    ----------
+    system_azimuth : float
+        direction of normal to slope on horizontal [radians]
+    system_zenith : float
+        tilt of normal to slope relative to vertical [radians]
+    tracker_azimuth : float
+        direction of tracker axes on horizontal [radians]
+
+    Returns
+    -------
+    tracker_zenith : float
+        tilt of tracker [radians]
+
+    Solving for the tracker tilt on a slope is derived in the following steps:
+
+    1. the trackers axes are in the system plane, so the ``z-coord = 0``
+
+    2. rotate the trackers ``[x_tr_sys, y_tr_sys, 0]`` back to the global, but
+       rotated by the tracker global azimuth if there is one, so that the
+       tracker axis is constrained to y-z plane so that ``x-coord = 0`` ::
+
+        Rx_sys = [[1,           0,            0],
+                  [0, cos(sys_ze), -sin(sys_ze)],
+                  [0, sin(sys_ze),  cos(sys_ze)]]
+
+        Rz_sys = [[cos(sys_az-tr_az), -sin(sys_az-tr_az), 0],
+                  [sin(sys_az-tr_az),  cos(sys_az-tr_az), 0],
+                  [                0,                  0, 1]]
+
+        tr_rot_glo = Rz_sys.T * (Rx_sys.T * [x_tr_sys, y_tr_sys, 0])
+
+        tr_rot_glo = [
+          [ x_tr_sys*cos(sys_az-tr_az)+y_tr_sys*sin(sys_az-tr_az)*cos(sys_ze)],
+          [-x_tr_sys*sin(sys_az-tr_az)+y_tr_sys*cos(sys_az-tr_az)*cos(sys_ze)],
+          [                                             -y_tr_sys*sin(sys_ze)]]
+
+    3. solve for ``x_tr_sys`` ::
+
+        x_tr_sys*cos(sys_az-tr_az)+y_tr_sys*sin(sys_az-tr_az)*cos(sys_ze) = 0
+        x_tr_sys = -y_tr_sys*tan(sys_az-tr_az)*cos(sys_ze)
+
+    4. so tracker axis tilt, ``tr_ze = arctan2(tr_rot_glo_z, tr_rot_glo_y)`` ::
+
+        tr_rot_glo_y = y_tr_sys*cos(sys_ze)*(
+          tan(sys_az-tr_az)*sin(sys_az-tr_az) + cos(sys_az-tr_az))
+
+        tan(tr_ze) = -y_tr_sys*sin(sys_ze) / tr_rot_glo_y
+
+    The trick is multiply top and bottom by cos(sys_az-tr_az) and remember that
+    ``sin^2 + cos^2 = 1`` (or just use sympy.simplify) ::
+
+        tan(tr_ze) = -tan(sys_ze)*cos(sys_az-tr_az)
+    """
+    sys_az_rel_to_tr_az = system_azimuth - tracker_azimuth
+    tan_tr_ze = -np.cos(sys_az_rel_to_tr_az) * np.tan(system_zenith)
+    return -np.arctan(tan_tr_ze)
+
+
+def calc_system_tracker_side_slope(
+        tracker_azimuth, tracker_zenith, system_azimuth, system_zenith):
+    """
+    Calculate the slope perpendicular to the tracker axis relative to the
+    system plane containing the axes as well as the rotation of the tracker
+    axes relative to the system plane. Note in order for the backtracking
+    algorithm to work correctly on a sloped system plane, the side slope must
+    be applied to the tracker rotation.
+
```

Comment:
I think the system plane needs to be formally defined somewhere in the documentation of a public function. The other functions can reference that definition. 

Suggested text: The system plane is defined by the tracker azimuth, tracker tilt and ground surface. The y-axis of the system plane is along the tracker axis in the direction of the tracker azimuth. The x-axis of the system plane is 90 degrees clockwise from the y-axis, and is parallel to the ground surface. The z-axis of the system plane is perpendicular to the plane formed by the x and y axes using the right-hand rule.

Reply:
>The fact that the descriptions of `system_azimuth` and `system_zenith` both refer to "slope" but not "system" might imply that they should be named `slope_azimuth` and `slope_zenith` [or `slope_tilt`].

I'm starting to agree with this. "System" is misleading, however, "slope" could be misleading too, if the pile heights vary. I've used "system" because it's really the plane that the trackers are in that's important, but in this context maybe "slope" is better?

>Suggested text: The system plane is defined by the tracker azimuth, tracker tilt and ground surface. The y-axis of the system plane is along the tracker axis in the direction of the tracker azimuth. The x-axis of the system plane is 90 degrees clockwise from the y-axis, and is parallel to the ground surface. The z-axis of the system plane is perpendicular to the plane formed by the x and y axes using the right-hand rule.

If I understand you correctly this would define a **system (or "slope") plane** that _was_ parallel to the tracker axes, whereas the point of this PR is define a system (or "slope") plane that is _not_ parallel to the tracker axes.

I wonder if the use of "system" was a poor choice. If I replace it for "slope" then for example NS trackers on an 10&deg; mostly east facing slope would have a `slope_azimuth` mostly due east, and a `slope_tilt` of 10&deg;. Then the corresponding "slope" plane y-axis would be mostly east (same direction as "slope" azimuth) and the "slope" plane x-axis in a RHS would be mostly south.

Here's an image of a system where the sloped plane of the trackers is facing slightly north of due east at a slope of about 10&deg;. The trackers are aligned NS in the global coordinate system, so the global YZ plane contains the tracker axes, and the global XY plane is horizontal. Like in the example above, the "slope" plane y-axis would follow its azimuth and face mostly east, so _not_parallel to the tracker axes which are NS.

![twisted_trackers](https://user-images.githubusercontent.com/1385621/74074933-36808a00-49c5-11ea-8850-c93d38abdfa6.png)

Now the problem becomes, how to derive the axis tilt and side slope to use in the existing `singleaxistracking` algorithm. That's what the algorithm in [`calc_tracker_axis_tilt()`](../../../mikofski/pvlib-python/blob/5dc2c5479e064540ba435ae149015eeb9925dd41/pvlib/tracking.py#L639-L677) tries to explain, but I must admit it's harder to follow in the ASCII only docstring, even though it's only been a month since my last commit. I believe Kevin Anderson derived similar equations, but maybe in a simpler more logical way. Perhaps he can do a better job than I have explaining? If my tracker PVSC paper is accepted, I will put the algorithms in there, pending checks for you, Kevin, and anyone else who want's to chip in. Thanks!

Reply:
Now I'm more confused. There's three planes at work here. The first is the ground surface. The third is the time-varying plane defined by the rotating modules. The second is still vague. If all trackers have the same azimuth and tilt and height above ground, and there's more than one tracker, then there is a plane defined by the set of vectors along the tracker azimuths. I thought this was the "system" plane. But if there's only one tracker, the plane can't be defined without specifying the not-azimuth, not-zenith vector. I thought that vector would parallel the sloped ground.

Reply:
I agree with Cliff -- I've been interpreting "system plane" as the plane containing the array's tracker axes.  Putting aside the complication @mikofski mentioned elsewhere about nonplanar slopes, I'd say that the point of this PR is not to distinguish the plane of the slope from the plane of the system (which are parallel) but rather to allow a distinction between the direction of falling slope and the azimuth of the trackers.  Put another way, the X-Y plane of the slope plane is parallel to the X-Y plane of the system, but this PR introduces a rotation around the Z axis.

Reply:
"slope plane" is different from the ground surface plane?

Reply:
>[KA] I agree with Cliff -- I've been interpreting "system plane" as the plane containing the array's tracker axes.

We all agree. This is what I have attempted to define the "system" (or "slope") plane to be. The plane that contains all of the tracker axes.

>[CH] Now I'm more confused. There's three planes at work here. The first is the ground surface. The second is the time-varying plane defined by the rotating modules.

Sorry, I'm not doing a very good job here. Yes, I think there are 3 surfaces:
1. The "ground" (which should not matter),
2. the "system" (or "slope"), and
3. the PV surface of the tracker which is time varying.

>[CH] "slope plane" is different from the ground surface plane?

Yes, I think this may be where the confusion lies for me and maybe others? The ground might not be the same as the "system" (or "slope") plane containing the tracker axes. For example, here's the terrain at the [Hopewell Friends Solar project in NC](https://goo.gl/maps/xkqZoCVRMtQypWJA9) ([DOE DE-FOA-0001840 CX101337 "Capturing the full benefits of bifacial modules to achieve an LCOE of 3c/kWh through a regional optimization of the electrical architecture"](https://www.energy.gov/sites/prod/files/2019/02/f59/101337_0.pdf)), which has varying slope in 2 directions:

![Hopewell_Civil_Base](https://user-images.githubusercontent.com/1385621/74183214-61532400-4bf9-11ea-85dc-f8436cbebad7.png)

>[KA] Putting aside the complication @mikofski mentioned elsewhere about nonplanar slopes,

Yes, I agree, let's put this aside for right now, because in principle the ground should not matter.

>[CH] The third plane, which I assumed is the "system" plane, contains the vectors along the tracker axes, which is not a complete definition if there's only one tracker. If all trackers have the same azimuth and tilt and height above ground, and there's more than one tracker, then there is a plane defined by the set of vectors along the tracker azimuths. I thought this was the "system" plane. 

Yes, I agree, this is what I have defined the "system" (or "slope") plane to be. So maybe it's me who is not understanding. Can you walk me through this suggestion?
>[CH] Suggested text: The system plane is defined by the tracker azimuth, tracker tilt and ground surface. The y-axis of the system plane is along the tracker axis in the direction of the tracker azimuth. The x-axis of the system plane is 90 degrees clockwise from the y-axis, and is parallel to the ground surface. The z-axis of the system plane is perpendicular to the plane formed by the x and y axes using the right-hand rule.

If "The y-axis of the system plane is along the tracker axis in the direction of the tracker azimuth" then this constrains the problem to only one "system" (or "slope") plane which is already solved by the current implementation as far as I understand it. I assume, perhaps incorrectly that "The y-axis of the system plane" means the direction that the "system" (or "slope") plane azimuth is pointing, but the problem that I am trying to solve is when that is **not** parallel to the tracker axes, but pointing a different direction.

If "The x-axis of the system plane is 90 degrees clockwise from the y-axis, and is parallel to the ground surface." then this also implies to me the situation already solved for by the current implementation. It was these comments that confused me. Sorry.

Also this statement also confuses me, sorry:
>[CH] But if there's only one tracker, the plane is defined by the tracker azimuth and the not-azimuth, not-zenith vector. I thought that this vector would parallel the sloped ground.

Why only one tracker? This PR involves at least two trackers to define the "system" (or "slope") plane that contains all of the trackers. And if there 3 or more trackers, then they would have to be constrained to this "system" (or "slope") plane as well. I think we are all saying the same thing now right?

>[KA] I'd say that the point of this PR is not to distinguish the plane of the slope from the plane of the system (which are parallel) **but rather to allow a distinction between the direction of falling slope and the azimuth of the trackers [axes]**.

Yes, I think this is what I'm trying to convey with this PR. I added emphasis, and also the word "axes" to tracker to clarify the further distinction between the tracker axes azimuth and the azimuths of the tracker surfaces (the 3rd plane).

>[KA] Put another way, the X-Y plane of the slope plane is parallel to the X-Y plane of the system, but this PR introduces a rotation around the Z axis.

**Yes, I think that's it! Does this work for everyone?**

Just one tiny clarification, the reason why "this PR introduces a rotation around the Z axis" is to allow tracking on slopes that are not parallel with the tracker axes. EG: a slope facing east, but tracker axes that run NS. That's it.

_longest comment ever_